tablets: Make load balancing capacity-aware

tgrabiec opened the pull request:
Before this patch, the load balancer was equalizing tablet count per
shard, so it achieved balance assuming that:
 1) tablets have the same size
 2) shards have the same capacity

That can cause imbalance of utilization if shards have different
capacity, which can happen in heterogeneous clusters with different
instance types. One of the causes for capacity difference is that
larger instances run with fewer shards due to vCPUs being dedicated to
IRQ handling. This makes those shards have more disk capacity, and
more CPU power.

After this patch, the load balancer equalizes shard's storage
utilization, so it no longer assumes that shards have the same
capacity. It still assumes that each tablet has equal size. So it's a
middle step towards full size-aware balancing.

One consequence is that to be able to balance, the load balancer need
to know about every node's capacity, which is collected with the same
RPC which collects load_stats for average tablet size. This is not a
significant set back because migrations cannot proceed anyway if nodes
are down due to barriers. We could make intra-node migration
scheduling work without capacity information, but it's pointless due
to above, so not implemented.

Also, per-shard goal for tablet count is still the same for all nodes in the cluster,
so nodes with less capacity will be below limit and nodes with more capacity will
be slightly above limit. This shouldn't be a significant problem in practice, we could
compensate for this by increasing the limit.

Refs #23042

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -2714,6 +2786,17 @@ class load_balancer {
+                continue;
+            }
+            if (!node.capacity) {
+                lblogger.info("Cannot balance because capacity of node {} (or more) is unknown", host);
```
Need to fall back to capacity = 1 if TABLET_LOAD_STATS_V2 is not enabled (rolling upgrade).

nodep commented on service/tablet_allocator.cc:
```diff
@@ -1542,25 +1583,35 @@ class load_balancer {
         // Load per shard in a perfectly balanced state.
-        // Assumes each shard has same capacity.
-        double desired_load_per_shard = double(total_load) / total_shard_count;
+        double desired_load_per_shard = total_load / _total_capacity_storage;
```
The desired lode calculated here is the valid for shard, node or dc. Maybe we should name it `desired_load_for_table`

tgrabiec replied:
I'll change it to `ideal_table_load`.

tgrabiec replied:
It's the same per shard/node/dc.

tgrabiec commented:
In V2:
  - fixed unit tests and perf_load_balancing tool
  - fixed crashes due to attempt to access disengaged `node_load::capacity`
      - Replaced on_internal_error() in plan making in case capacity is not known with returning optional<> so that calling code is aware to be careful with capacity - smaller chance for hitting on_internal_error()
  - stats refreshing now has to tolerate a node down, at least for refreshing capacity (table stats are still not refreshed)
  - trigger stats refreshing after bootstrapping a node so that load balancing is not delayed by stats refresh period
  - dropped "tolerance" when calculating badness of migration, it's not meaningful with capacity-aware load, and we don't seem to need it with this metric.
  - fixed confusing log message when topology coordinator sees that load balancer's plan is empty (it shouldn't say "tablets are balanced")
  - fall back to assuming equal capacity when tablet_load_stats_v2 cluster feature is not enabled

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15674/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15674/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_custom/test_tablets_removenode
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3726/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6842/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14865/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15674/)
#### Build Details:
- Duration: 6 hr 39 min
- Builder: spider2.cloudius-systems.com

bhalevy commented on idl/storage_service.idl.hh:
```diff
@@ -24,10 +24,15 @@ struct table_load_stats final {
 
+struct load_stats {
+    std::unordered_map<::table_id, locator::table_load_stats> tables;
+    std::unordered_map<locator::host_id, uint64_t> capacity;
```
Can't we just append an optional member for backward compatibility?
Why do we need a new verb?

tgrabiec replied:
It's mentioned in the commit description: the object is marked final in IDL.

avikivity replied:
Why a map? Isn't the RPC issued to a single node?

avikivity replied:
I guess it's because of load_stats operator+=. But there's no need for the rpc payload to be the same as the accumulator. We could have global_load_stats(host_id, load_stats) to transform.

bhalevy commented on db/config.cc:
```diff
@@ -542,6 +542,8 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , data_file_directories(this, "data_file_directories", "datadir", value_status::Used, { },
         "The directory location where table data (SSTables) is stored.")
+    , data_file_capacity(this, "data_file_capacity", liveness::LiveUpdate, value_status::Used, 0,
+        "Total capacity in bytes for storing data files. If not set, will use file system's capacity.")
```
Since the effect is limited to tablets, let's document that in the help string.

bhalevy commented on utils/disk_space_monitor.cc:
```diff
@@ -76,7 +76,11 @@ future<> disk_space_monitor::poll() {
-    return engine().file_system_space(_data_dir.native());
+    auto space = co_await engine().file_system_space(_data_dir.native());
+    if (_cfg.capacity_override()) {
+        space.capacity = _cfg.capacity_override();
```
But now `free` and `available` are out of sync with capacity.
Maybe they need to be adjusted by (space.capacity - capacity_override) since this space is neither free nor available.

Also, should we allow capacity_override to be greater than the filesystem capacity?

tgrabiec replied:
> But now `free` and `available` are out of sync with capacity. Maybe they need to be adjusted by (space.capacity - capacity_override) since this space is neither free nor available.

Right, will fix.

> 
> Also, should we allow capacity_override to be greater than the filesystem capacity?

Since the setting is for manual override and emergency situations, I think it's better to leave it unconstrained in case we need to compensate for bugs which would otherwise under-allocate the node.

avikivity replied:
Agree

bhalevy commented on test/perf/tablet_load_balancing.cc:
```diff
@@ -374,7 +364,7 @@ future<results> test_load_balancing_with_many_tables(params p, bool tablet_aware
         check_balance();
 
-        rebalance_tablets(e.get_tablet_allocator().local(), stm);
+        rebalance_tablets(e);
```
This patch can be sent (and backported) independently of the series, no?

tgrabiec replied:
Yes

bhalevy commented on test/topology_custom/test_tablets_removenode.py:
```diff
@@ -29,6 +29,23 @@ async def run_async_cl_all(cql, query: str):
+
+    logger.info('Removing a node')
+    await manager.server_stop(servers[2].server_id)
+    await manager.server_restart(servers[0].server_id) # topo coordinator
```
Is it guaranteed that the topology coordinator will run on servers[0]?
Is it because it bootstrapped first?
Maybe it'd be more robust to look for it dynamically.

tgrabiec replied:
Will change to discover dynamically.

tgrabiec replied:
> Is it guaranteed that the topology coordinator will run on servers[0]? Is it because it bootstrapped first? 

Yes, and it will step down if quorum is lost. That shouldn't happen here, unless nodes are not responding for some reason (e.g. reactor stall). Better to be robust against this, yes.

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -432,14 +442,41 @@ class load_balancer {
+
+        // Result engaged for !drained nodes.
+        std::optional<load_type> shard_load(shard_id shard, uint64_t shard_tablet_count, uint64_t target_tablet_size) const {
+            if (!capacity) {
```
I guess that the compiler may very well optimize this, but we could call `capacity_per_shard` and check its result rather than double checking `capacity` at each level.

tgrabiec replied:
Ok

bhalevy commented:
@avikivity can you please review this series?

avikivity commented:
> migrations cannot proceed anyway if nodes
are down due to barriers.

@mykaul we must work to remove this limitation, or a dead node/rack/datacenter freezes the cluster

> We could make intra-node migration
scheduling work without capacity information, but it's pointless due
to above, so not implemented.

Is it really needed? To migrate between two nodes, both must be up. Anything on dead node doesn't change the balance.

avikivity commented on utils/disk_space_monitor.hh:
```diff
@@ -39,6 +39,7 @@ public:
         updateable_value<int> high_polling_interval;
         // Use high_polling_interval above this threshold
         updateable_value<float> polling_interval_threshold;
+        updateable_value<uint64_t> capacity_override; // 0 means no override.
```
Unfortunate to mix control and data

avikivity commented on service/topology_coordinator.cc:
```diff
@@ -2997,7 +2999,15 @@ future<> topology_coordinator::refresh_tablet_load_stats() {
+                if (_load_stats_per_node.contains(dst)) {
+                    node_stats = _load_stats_per_node[dst];
+                } else {
+                    rtlogger.warn("raft topology: Unable to refresh table load status because node {} is down.", dst);
```
This will be too noisy is a node is down for half an hour of maintenance.

avikivity commented on service/topology_coordinator.cc:
```diff
@@ -3048,6 +3059,11 @@ future<> topology_coordinator::refresh_tablet_load_stats() {
     }
+
+    if (table_load_stats_invalid) {
+        stats.tables.clear();
```
Why are they invalid if one node is down? The rest can still be balanced.

tgrabiec replied:
They are invalid because of the way we calculate the stats. We divide total size by RF. We could change it, but it's out of scope.

avikivity commented on service/topology_coordinator.cc:
```diff
@@ -2295,6 +2295,10 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                                                    "bootstrap: read fence completed");
+                    // Make sure the load balancer knows the capacity for the new node immediately.
+                    (void)_tablet_load_stats_refresh.trigger().handle_exception([] (auto ep) {
+                        rtlogger.warn("Error during tablet load stats refresh: {}", ep);
```
Ignoring a future needs a comment explaining why it is safe. Is it in fact safe?

Better, signal a fiber to do this (out of scope).

tgrabiec replied:
> Ignoring a future needs a comment explaining why it is safe. Is it in fact safe?
> 
> Better, signal a fiber to do this (out of scope).

It's safe, we join() with _tablet_load_stats_refresh in topology_coordinator::close(), which waits for all triggered jobs.

avikivity commented on service/tablet_allocator.cc:
```diff
@@ -346,8 +346,10 @@ class load_balancer {
+    // Represents metric for load which we want to equalize between shards or nodes.
+    // Load balancer equalizes storage utilization.
+    // It is assumed that each tablet has equal size and that shards and nodes can have different capacity.
+    // So we equalize: tablet_count * target_tablet_size / capacity_in_bytes.
```
We can consider a metric that mixes absolute free space and relative free space, not just relative free space as here.

Relative free space makes sense when considering that a shard with twice the capacity will see twice the ingest rate.

Absolute free space makes sense when considering non-proportional overhead such as group0 work.

But out of scope for now, relative free space is fine for i4i, maybe not for things like m7gd.

tgrabiec replied:
Why do you think group0 work (which work?) is relevant for balancing?

avikivity commented on service/tablet_allocator.cc:
```diff
@@ -419,6 +422,13 @@ class load_balancer {
+            if (!capacity) {
+                return std::nullopt;
+            }
+            return double(*capacity) / shard_count;
```
Could write: `return capacity.transform([] (double c) { return c / shard_count; });`. Would be more impressive with less verbose lambdas.

tgrabiec replied:
Yes, monads :heart_eyes:

avikivity commented:
Looks good, but need (possibly simulated) tests for heterogeneous nodes.

I want the wire protocol tightened not to include the node.

@nodep please review as it impacts your follow-on work

tgrabiec commented:
> > migrations cannot proceed anyway if nodes
> > are down due to barriers.
> 
> @mykaul we must work to remove this limitation, or a dead node/rack/datacenter freezes the cluster
> 
> > We could make intra-node migration
> > scheduling work without capacity information, but it's pointless due
> > to above, so not implemented.
> 
> Is it really needed? To migrate between two nodes, both must be up. Anything on dead node doesn't change the balance.

That remark was about intra-node balancing (within one node) of an UP node. We could balance the node even if we don't have capacity info for other nodes.

cezarmoise commented:
> Looks good, but need (possibly simulated) tests for heterogeneous nodes.



https://argus.scylladb.com/tests/scylla-cluster-tests/39c16c6e-770e-4199-b996-8de0d97a9d52
![tomek](https://github.com/user-attachments/assets/66db4e89-80ca-47eb-a552-7f573f21c16e)

compare to before
![tomek-old](https://github.com/user-attachments/assets/19f9f867-aea1-48b0-9acb-9413be521912)

tgrabiec commented:
Blocked by https://github.com/scylladb/scylla-dtest/pull/5599

bhalevy commented:
> Blocked by [scylladb/scylla-dtest#5599](https://github.com/scylladb/scylla-dtest/pull/5599)

Qd

tgrabiec commented:
In V3:
  - lowered log level to debug for "raft topology: Unable to refresh table load on {} because it's down."
  - Improved reliability against leader failover in test_removenode_with_coordinator_restart by checking which node is the leader
    - added test doc
  - Avoided double-checking of `node_load::capacity` at different levels, made get_avg_load() return optional<>:
     ```
    @@ service/tablet_allocator.cc: class load_balancer {
              bool drained = false;
              const locator::node* node; // never nullptr
      
    +         // The average shard load on this node.
    ++        // Valid only when "capacity" is set.
    +         load_type avg_load = 0;
    + 
    +         absl::flat_hash_map<table_id, size_t> tablet_count_per_table;
     @@ service/tablet_allocator.cc: class load_balancer {
      
              utils::chunked_vector<skipped_candidate> skipped_candidates;
    @@ service/tablet_allocator.cc: class load_balancer {
                  return node->dc_rack().dc;
              }
     @@ service/tablet_allocator.cc: class load_balancer {
    +             return node->get_state();
              }
      
    -         // Call when tablet_count changes.
    +-        // Call when tablet_count changes.
     -        void update() {
     -            avg_load = get_avg_load(tablet_count);
    ++        // Call when tablet_count or capacity changes.
     +        void update(uint64_t target_tablet_size) {
    -+            if (capacity) {
    -+                avg_load = *get_avg_load(tablet_count, target_tablet_size);
    ++            if (auto load = get_avg_load(tablet_count, target_tablet_size)) {
    ++                avg_load = *load;
     +            }
    -+        }
    -+
    -+        // Result engaged when !drained.
    -+        std::optional<load_type> get_avg_load(uint64_t tablets, uint64_t target_tablet_size) const {
    -+            if (!capacity) {
    -+                return std::nullopt;
    -+            }
    -+            return double(tablets) * target_tablet_size / *capacity;
              }
      
     -        load_type get_avg_load(uint64_t tablets) const {
    ++        // Result engaged when !drained.
    ++        std::optional<load_type> get_avg_load(uint64_t tablets, uint64_t target_tablet_size) const {
    ++            return capacity.transform([&](uint64_t capacity) {
    ++                return load_type(tablets * target_tablet_size) / capacity;
    ++            });
    ++        }
    ++
     +        double tablets_per_shard(uint64_t tablets) const {
                  return double(tablets) / shard_count;
              }
    @@ service/tablet_allocator.cc: class load_balancer {
     +
     +        // Result engaged for !drained nodes.
     +        std::optional<load_type> shard_load(shard_id shard, uint64_t shard_tablet_count, uint64_t target_tablet_size) const {
    -+            if (!capacity) {
    ++            auto cap = capacity_per_shard();
    ++            if (!cap) {
     +                return std::nullopt;
     +            }
    -+            return double(shard_tablet_count) * target_tablet_size / *capacity_per_shard();
    ++            return double(shard_tablet_count) * target_tablet_size / *cap;
     +        }
     +
              auto shards_by_load_cmp() {

     ``` 
  - Add boost unit test for balancing of heterogeneous cluster: test_balancing_heterogeneous_cluster
  - disk_space_monitor:
     - adjust available and free space when capacity override is in force
     - trigger refresh of space info immediately when config changes
     - added unit test for capacity override

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15794/)
#### Build Failure:
```
[2025-03-06T12:02:06.927Z] FAILED: build/dev/utils/disk_space_monitor.hh.o 
[2025-03-06T12:02:06.927Z] clang++ -MD -MT build/dev/utils/disk_space_monitor.hh.o -MF build/dev/utils/disk_space_monitor.hh.o.d -std=gnu++23 -I/jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include -I/jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/gen/include -Werror=unused-result -fstack-clash-protection -DSEASTAR_API_LEVEL=7 -DSEASTAR_BUILD_SHARED_LIBS -DSEASTAR_SSTRING -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSEASTAR_LOGGER_COMPILE_TIME_FMT -DSEASTAR_SCHEDULING_GROUPS_COUNT=19 -DSEASTAR_LOGGER_TYPE_STDOUT -DSEASTAR_TYPE_ERASE_MORE -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_THREAD_NO_LIB -DBOOST_THREAD_DYN_LINK -DFMT_SHARED -I/usr/include/p11-kit-1 -std=gnu++23 -g -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -DSEASTAR_NO_EXCEPTION_HACK -march=westmere -DBOOST_ALL_DYN_LINK    -fvisibility=hidden -Ikmipc/kmipc-2.1.0t-rhel84_64/include -DHAVE_KMIP -isystem abseil -Wall -Werror -Wextra -Wimplicit-fallthrough -Wno-mismatched-tags -Wno-c++11-narrowing -Wno-overloaded-virtual -Wno-unused-parameter -Wno-unsupported-friend -Wno-missing-field-initializers -Wno-deprecated-copy -Wno-enum-constexpr-conversion -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -g -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -DSEASTAR_NO_EXCEPTION_HACK -march=westmere -DDEVEL -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSCYLLA_ENABLE_ERROR_INJECTION -DSCYLLA_ENABLE_PREEMPTION_SOURCE -O2 -DSCYLLA_BUILD_MODE=dev -Xclang -fexperimental-assignment-tracking=disabled -iquote. -iquote build/dev/gen  --include utils/disk_space_monitor.hh -c -o build/dev/utils/disk_space_monitor.hh.o build/dev/gen/empty.cc
[2025-03-06T12:02:06.928Z] In file included from <built-in>:1:
[2025-03-06T12:02:06.928Z] ./utils/disk_space_monitor.hh:51:5: error: unknown type name condition_variable
[2025-03-06T12:02:06.928Z]    51 |     condition_variable _poll_cv;
[2025-03-06T12:02:06.929Z]       |     ^
[2025-03-06T12:02:06.929Z] 1 error generated.
```
#### Build Details:
- Duration: 36 min
- Builder: spider2.cloudius-systems.com

bhalevy commented on locator/load_sketch.hh:
```diff
@@ -236,6 +236,19 @@ public:
+            return std::nullopt;
+        }
+        auto capacity = stats.capacity.at(node);
+        return capacity > 0 ? double(n.load() * target_tablet_size) / capacity : 0;
```
This implementation is searching node twice each time.
It's a bit less readable but more efficient to use find, like:
```c++
        auto nodes_it = _nodes.find(node);
        if (nodes_it == _nodes.end()) {
            return std::nullopt;
        }
        auto& n = nodes_it.second;
        auto stats_it = stats.capacity.find(node);
        if (stats_it == stats.capacity.end()) {
            return std::nullopt;
        }
        auto capacity = stats_it->second;
        return capacity > 0 ? double(n.load() * target_tablet_size) / capacity : 0;
```

tgrabiec replied:
Efficiency doesn't matter here, so simplicity wins.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15802/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15802/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/database_test
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: cluster/test_tablets_removenode
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14977/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6951/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3836/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15802/)
#### Failed Tests (1/45113):
- [test_topology_ops_encrypted[True].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15802/testReport/junit/cluster/test_topology_ops_encrypted/Tests___Unit_Tests___test_topology_ops_encrypted_True__debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_ops_encrypted[True].debug.3)
#### Build Details:
- Duration: 10 hr
- Builder: i-0656cfb90b291442d (m5ad.12xlarge)

cezarmoise commented:
@tgrabiec 

With latest changes, got out of space on the bigger instances, which has never happened before.
[Argus](https://argus.scylladb.com/tests/scylla-cluster-tests/a7759eae-9f66-4987-949b-614b38d48026) [Monitoring](http://3.235.152.206:3000/)
Initial cluster 3 x 2XL, write to 90%, add 3 L, write to 90% again. Initial tablets 1024.

Disk usage
![Screenshot 2025-03-07 115425](https://github.com/user-attachments/assets/175ec972-1f96-45b1-90e6-840be2d37d4a)

Tablets per shard from the 1st write to 90%
![Screenshot 2025-03-07 115834](https://github.com/user-attachments/assets/9530c3c4-4219-4a28-8144-d21d09a51e48)

Tablets per shard from the 2nd write to 90%
![Screenshot 2025-03-07 115859](https://github.com/user-attachments/assets/52196b44-1985-4b8f-8370-b10edc38dffd)

tgrabiec commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15802/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15802/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/database_test 🔹 boost/tablets_test 🔹 cluster/test_tablets_removenode ✅ - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14977/) ✅ - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6951/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3836/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15802/)
> 
> #### Failed Tests (1/45113):
> * [test_topology_ops_encrypted[True].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15802/testReport/junit/cluster/test_topology_ops_encrypted/Tests___Unit_Tests___test_topology_ops_encrypted_True__debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_ops_encrypted%5BTrue%5D.debug.3)
> 
> #### Build Details:
> * Duration: 10 hr
> * Builder: i-0656cfb90b291442d (m5ad.12xlarge)

https://github.com/scylladb/scylladb/issues/22809

tgrabiec commented:
> @tgrabiec
> 
> With latest changes, got out of space on the bigger instances, which has never happened before. [Argus](https://argus.scylladb.com/tests/scylla-cluster-tests/a7759eae-9f66-4987-949b-614b38d48026) [Monitoring](http://3.235.152.206:3000/) Initial cluster 3 x 2XL, write to 90%, add 3 L, write to 90% again. Initial tablets 1024.

Did initial_tablets change in this test?

I would try without initial_tablets on the second table. I can see that none of the two tables grow beyond that count, so it could be the cause of imbalance - smaller table has the same tablet count as the larger table, which amplifies https://github.com/scylladb/scylladb/issues/22982. Lack of size-based balancing is also a factor.

I can see that "scylla_load_balancer_load" (allocated disk utilization) is balanced:
![Screenshot from 2025-03-07 13-03-36](https://github.com/user-attachments/assets/16ebdbce-aff9-4713-90be-a3c1a72df1a0)

tgrabiec commented:
> > @tgrabiec
> > With latest changes, got out of space on the bigger instances, which has never happened before. [Argus](https://argus.scylladb.com/tests/scylla-cluster-tests/a7759eae-9f66-4987-949b-614b38d48026) [Monitoring](http://3.235.152.206:3000/) Initial cluster 3 x 2XL, write to 90%, add 3 L, write to 90% again. Initial tablets 1024.
> 
> Did initial_tablets change in this test?
> 
> I would try without initial_tablets on the second table. I can see that none of the two tables grow beyond that count, so it could be the cause of imbalance - smaller table has the same tablet count as the larger table, which amplifies #22982. Lack of size-based balancing is also a factor.
> 
> I can see that "scylla_load_balancer_load" (allocated disk utilization) is balanced: ![Screenshot from 2025-03-07 13-03-36](https://private-user-images.githubusercontent.com/283695/420336269-16ebdbce-aff9-4713-90be-a3c1a72df1a0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEzNTA5MzksIm5iZiI6MTc0MTM1MDYzOSwicGF0aCI6Ii8yODM2OTUvNDIwMzM2MjY5LTE2ZWJkYmNlLWFmZjktNDcxMy05MGJlLWEzYzFhNzJkZjFhMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMwN1QxMjMwMzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yOTNjNjZlZWJjNTFjNDNjYmIyOGUxZjQ4MGVlNjdlNzQwNzFhOGQwODc3MTNjYmVjMjMxNGQ3M2YwY2NjYzhiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.V3xTcBfmYo_dFJbMn7okG79q1HNl19d_v_3bBJbUo9M)

So the difference must be in actual tablet sizes on disk plus per-table imbalance.

tgrabiec commented:
@cezarmoise Setting initial tablet count per keyspace is not recommended with current master. Please drop it. For speeding up ingestion, use per-table expected table size hint.

avikivity commented:
Will wait for test results.

bhalevy commented:
> Will wait for test results.

@avikivity what kind of test results are you waiting for?
FWIW, CI already passed (https://github.com/scylladb/scylladb/pull/23079#issuecomment-2707508446)

cezarmoise commented:
> @cezarmoise Setting initial tablet count per keyspace is not recommended with current master. Please drop it. For speeding up ingestion, use per-table expected table size hint.

Re-ran both latest version and previous version with `tablets = {'min_per_shard_tablet_count': 10}`.
Did not write to 90% this time to speed it up.
The issue did not replicate.

---
Scylla version: 2025.2.0~dev-20250304.3491311ea6b2 ( from first try)
[Argsu](https://argus.scylladb.com/tests/scylla-cluster-tests/fe1903b1-03c1-4d74-8158-b9af2b738e87) [Monitoring](http://44.204.220.13:3000/d/OS-master/os-metrics?orgId=1)
Balanced correctly on disk used %
![Screenshot 2025-03-10 123948](https://github.com/user-attachments/assets/edd76598-d1d1-4416-ac3c-c17777f92c2d)

---
Scylla version: 2025.2.0~dev-20250306.c4714180ccc4 (from 2nd try)
[Argus](https://argus.scylladb.com/tests/scylla-cluster-tests/3f89d5f7-c505-4df8-a244-945b14dec2e6) [Monitoring](http://44.220.70.8:3000/d/OS-master/os-metrics?orgId=1)
Balanced correctly on disk used %
![Screenshot 2025-03-10 124110](https://github.com/user-attachments/assets/030a5ee0-9491-40b7-99fd-5d7547dcbd3b)

---

I will try to re-run the exact scenario that caused generated that issue and update when it's done.

cezarmoise commented:
Same scenario with 90% utilization is also ok
[Argus](https://argus.scylladb.com/tests/scylla-cluster-tests/0f9b962b-b399-437e-b983-52fcb604a87a) [Monitor](http://35.153.49.227:3000/)
![Untitled](https://github.com/user-attachments/assets/1993b7c8-d20a-4ff4-9c7c-4ae048d8f444)

avikivity commented:
> > Will wait for test results.
> 
> @avikivity what kind of test results are you waiting for? FWIW, CI already passed ([#23079 (comment)](https://github.com/scylladb/scylladb/pull/23079#issuecomment-2707508446))

Update of https://github.com/scylladb/scylladb/pull/23079#issuecomment-2706026754

avikivity commented:
> > > Will wait for test results.
> > 
> > 
> > @avikivity what kind of test results are you waiting for? FWIW, CI already passed ([#23079 (comment)](https://github.com/scylladb/scylladb/pull/23079#issuecomment-2707508446))
> 
> Update of [#23079 (comment)](https://github.com/scylladb/scylladb/pull/23079#issuecomment-2706026754)

Which was, in fact, updated today.

cezarmoise commented:
with `tablets = {min_per_shard_tablet_count: 128}` (which also leads to 1024 tablets) it's also ok
[Argus](https://argus.scylladb.com/tests/scylla-cluster-tests/79bf63dc-9499-434c-a774-1e557b40b0e4) [Monitor](http://34.240.240.30:3000/)
![Screenshot 2025-03-11 200008](https://github.com/user-attachments/assets/aa93e6c8-1322-4676-b1d5-20617e586bd0)

cezarmoise commented:
> > > Which was, in fact, updated today.
> > 
> > 
> > @avikivity
> > With `tablets = {initial: 1024}` it still happens [Argus](https://argus.scylladb.com/tests/scylla-cluster-tests/6bd21cf2-c586-42dd-91ea-f873bd5c5c31)
> 
> But from what I understood, it's not a regression, right? It's a new configuration which wasn't tested before.

Yes, I re-used a test that I knew would get out of space with old balancing.

cezarmoise commented:
> > > Which was, in fact, updated today.
> > 
> > 
> > @avikivity
> > With `tablets = {initial: 1024}` it still happens [Argus](https://argus.scylladb.com/tests/scylla-cluster-tests/6bd21cf2-c586-42dd-91ea-f873bd5c5c31) [Monitor](http://52.209.114.125:3000/) ![Screenshot 2025-03-11 153331](https://private-user-images.githubusercontent.com/1081930/421393628-b5db6cca-deb2-4005-9686-5d626126c624.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE3MDA4MTgsIm5iZiI6MTc0MTcwMDUxOCwicGF0aCI6Ii8xMDgxOTMwLzQyMTM5MzYyOC1iNWRiNmNjYS1kZWIyLTQwMDUtOTY4Ni01ZDYyNjEyNmM2MjQuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMTFUMTM0MTU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OTg0NDk0YWY2MDA5OGQ5YzJjM2I5NDRiMjY0MDdjNzQ1NWFhZDdhYzE4NzNmNjNjOGNhNDRjNTc2YzU4ZjkwZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.u6YhI1CG0_TFiAwC5TkkhIjlyd1ZDLzhK7doVPY0P1Y)
> > but not with `tablets = {min_per_shard_tablet_count: 10}` [Argus](https://argus.scylladb.com/tests/scylla-cluster-tests/99ae03f8-80c2-44de-b624-2eaa8673275d) [Monitor](http://54.217.10.174:3000/) ![Screenshot 2025-03-11 153511](https://private-user-images.githubusercontent.com/1081930/421394435-a7c02e82-3700-4cfe-929c-2ac8b3c9ad6d.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE3MDA4MTgsIm5iZiI6MTc0MTcwMDUxOCwicGF0aCI6Ii8xMDgxOTMwLzQyMTM5NDQzNS1hN2MwMmU4Mi0zNzAwLTRjZmUtOTI5Yy0yYWM4YjNjOWFkNmQuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMTFUMTM0MTU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWMwYjE4YjI5NDk4MjU5YjY4MTQ2Mjg1NzM2NGM3ZjFjNTkzOTE0NzhmZmVmNDZmYjQxNTQ5MTNiNDM2MjY2NiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Qv4aR1XOoWnoXG3x9WqoSdy6rtrZT79UcU6WTRcjFJk)
> > Both have the instances still up if needed.
> 
> I will investigate that.

If it helps, I also tried with initial 128, 256 and 512 and it does not replicate in those scenarios, it balances properly.

bhalevy commented:
> ⚠️ @tgrabiec PR body does not contain a Fixes reference to an issue and can not be backported
> 
> The following labels were removed:
> 
> * backport/2025.1

@tgrabiec what about the backport for this PR?
We want it in branch-2025.1 for 2025.1.1 right?

tgrabiec commented:
> > ⚠️ @tgrabiec PR body does not contain a Fixes reference to an issue and can not be backported
> > The following labels were removed:
> > 
> > * backport/2025.1
> 
> @tgrabiec what about the backport for this PR? We want it in branch-2025.1 for 2025.1.1 right?

Backport to 2025.1: https://github.com/scylladb/scylladb/pull/23443
